content/blog/2019/07/2019-07-03-jenkins-pipeline-stage-result-visualization-improvements.adoc

```diff
@@ -0,0 +1,74 @@
+
+In order to fix this issue, we introduced a new Pipeline API that can be used to attach additional result information to individual Pipeline steps. Visualization tools like Blue Ocean use this new API when deciding how a given stage should be displayed. Steps like `junit` that used to set only the overall build result now additionally use the new API to set step-level result information. We created the new `unstable` and `warnError` steps so that Pipeline authors with more complicated use cases can still take advantage of this new API.
+
+The core fixes for the issue are present in the following plugins, all of which require Jenkins 2.138.4 or newer:
+
+* Pipeline: API 2.34
+* Pipeline: Basic Steps 2.18 (requires a simultaneous update to Pipeline: Groovy 2.70)
+* Pipeline: Graph Analysis 1.10
+* Pipeline: Declarative 1.3.9
+* Blue Ocean 1.17.0
+
+Here is a screenshot from Blue Ocean of a Pipeline using the `unstable` step where only the failing stage is marked as unstable:
+
+image::/images/post-images/2019/07/2019-07-03-jenkins-pipeline-stage-result-visualization-improvements/unstable-stage-example.png[Visualization of a Pipeline in Blue Ocean with a single stage shown as unstable]
+
+== Examples
+
+Here are some examples of how to update your Pipelines to use the new improvements:
+
+* **Use the new link:https://jenkins.io/doc/pipeline/steps/workflow-basic-steps/#warnerror-catch-error-and-set-build-and-stage-result-to-unstable[`warnError`] step to catch errors and mark the build and stage as unstable.** `warnError` requires a single `String` parameter, which is a message to log when an error is caught. When `warnError` catches an error, it logs the message  and the error and sets the build and stage result to unstable. Using it looks like this:
++
+[source,groovy]
+----
+warnError('Script failed!') {
+  sh('false')
+}
+----
+
+* **Use the new link:https://jenkins.io/doc/pipeline/steps/workflow-basic-steps/#unstable-set-stage-result-to-unstable[`unstable`] step to set the build and stage result to unstable**. This step can be used as a direct replacement for `currentBuild.result = 'UNSTABLE'`, and may be useful in cases where `warnError` is not flexible enough. `unstable` requires a single `String` parameter, which is a message to log when the step runs. Using it might look like this:
++
+[source,groovy]
+----
+try {
+  sh('false')
+} catch (ex) {
+  unstable('Script failed!')
+}
+----
+
+* **link:https://plugins.jenkins.io/junit[JUnit Plugin]**: Update to version 1.28 of JUnit plugin or newer to pick up fixes for the `junit` step so that it correctly marks the stage as unstable.
```

Comment:
```suggestion
* **link:https://plugins.jenkins.io/junit[JUnit Plugin]**: Update to version 1.28 or newer to pick up fixes for the `junit` step so that it correctly marks the stage as unstable.
```

Redundant